concurrency/SortedRanges.java

```diff
@@ -0,0 +1,384 @@
+        if(endIndex<nextRange.getEndIndex()) {
+          //expand this range
+          endIndex = nextRange.getEndIndex();
+          break;
+        }
+      } else {
+        break;
+      }
+    }
+    add(startIndex,endIndex);
+  }
+  
+  /**
+   * Remove the range indices. If this range is  
+   * found in existing ranges, the existing ranges 
+   * are shrunk.
+   * If range is of 0 length, doesn't do anything.
+   * @param range Range to be removed.
+   */
+  synchronized void remove(Range range) {
+    if(range.isEmpty()) {
+      return;
+    }
+    long startIndex = range.getStartIndex();
+    long endIndex = range.getEndIndex();
+    //make sure that there are no overlapping ranges
+    SortedSet<Range> headSet = ranges.headSet(range);
+    if(headSet.size()>0) {
+      Range previousRange = headSet.last();
+      LOG.debug("previousRange "+previousRange);
+      if(startIndex<previousRange.getEndIndex()) {
+        //previousRange overlaps this range
+        //narrow down the previousRange
+        if(ranges.remove(previousRange)) {
+          indicesCount-=previousRange.getLength();
+          LOG.debug("removed previousRange "+previousRange);
+        }
+        add(previousRange.getStartIndex(), startIndex);
+        if(endIndex<=previousRange.getEndIndex()) {
+          add(endIndex, previousRange.getEndIndex());
+        }
+      }
+    }
+    
+    Iterator<Range> tailSetIt = ranges.tailSet(range).iterator();
+    while(tailSetIt.hasNext()) {
+      Range nextRange = tailSetIt.next();
+      LOG.debug("nextRange "+nextRange +"   startIndex:"+startIndex+
+          "  endIndex:"+endIndex);
+      if(endIndex>nextRange.getStartIndex()) {
+        //nextRange overlaps this range
+        //narrow down the nextRange
+        tailSetIt.remove();
+        indicesCount-=nextRange.getLength();
+        if(endIndex<nextRange.getEndIndex()) {
+          add(endIndex, nextRange.getEndIndex());
+          break;
+        }
+      } else {
+        break;
+      }
+    }
+  }
+  
+  private void add(long start, long end) {
+    if(end>start) {
+      Range recRange = new Range(start, end-start);
+      ranges.add(recRange);
+      indicesCount+=recRange.getLength();
+      LOG.debug("added "+recRange);
+    }
+  }
+  
+  public synchronized void readFields(DataInput in) throws IOException {
+    indicesCount = in.readLong();
+    ranges = new TreeSet<Range>();
+    int size = in.readInt();
+    for(int i=0;i<size;i++) {
+      Range range = new Range();
+      range.readFields(in);
+      ranges.add(range);
+    }
+  }
+
+  public synchronized void write(DataOutput out) throws IOException {
+    out.writeLong(indicesCount);
+    out.writeInt(ranges.size());
+    Iterator<Range> it = ranges.iterator();
+    while(it.hasNext()) {
+      Range range = it.next();
+      range.write(out);
+    }
+  }
+  
+  public String toString() {
+    StringBuffer sb = new StringBuffer();
+    Iterator<Range> it = ranges.iterator();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **211**. At this line the following field is not protected by synchronization: `ranges`. Your code typically protects this field with synchronization on object: `SortedRanges.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `SortedRanges.this`. For example, like at lines: **52**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **52**, **88**, **95**, **105**, **141**, **148**, **159**, **182**, **195**, **201**, and **202**.